substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -552,37 +554,34 @@ where
 			// The best reported block is in the pruned list. Report a new best block.
 			let is_in_pruned_list =
 				pruned_block_hashes.iter().any(|hash| *hash == current_best_block);
-			// The block is not the last finalized block.
+
+			// The best reported block is not the last finalized block.
 			//
 			// It can be either:
 			//  - a descendant of the last finalized block
 			//  - a block on a fork that will be pruned in the future.
 			//
 			// In those cases, we emit a new best block.
+			//
+			// When [`Self::generate_finalized_events`] reports the last finalized block as the best
+			// block, it also sets the `self.current_best_block`, if and only if the block
+			// was never reported as `NewBlock` before.
+			//
+			// Therefore, this condition is equivalent with
+			// `!std::matches(FollowEvent::BestBlockChanged(_), events.last())`.
+			// And that means we know for sure that at least a `NewBlock` was generated
+			// for the last finalized block.
 			let is_not_last_finalized = current_best_block != last_finalized;
 
 			if is_in_pruned_list || is_not_last_finalized {
```

Comment:
I don't get why you want to generate the `BestBlockChangedEvent` when the best block is not the last finalized? The best block could just be an descendant of the last finalized?

Reply:
Yep that makes sense, have checked instead that the best block is a descendant of the last finalized block 🙏